Make recorder .flv videos scrollable (#512)

oussamabadr opened the pull request:
I have written an implementation to fix the scrollable video using ffmpeg command, which will also reduce the video size as mentioned by @leonard84

I have also opened a related pull Request testcontainers/vnc-recorder#4 by adding **ffmpeg** to the docker image, which (I saw) necessary to avoid using another container and also simplify the fix.
Just to mention that some test won't pass untill the new **vnc-recorder** docker image is available with tag 1.2.0. 

(Thank to @kiview for his assitance during **hack-commit-push** (last june) which make contributing to **testcontainers** more clear for me.)

Fixes #512.

kiview commented on core/src/main/java/org/testcontainers/containers/VncRecordingContainer.java:
```diff
@@ -87,19 +88,27 @@ protected void configure() {
-        try(InputStream inputStream = streamRecording()) {
-            Files.copy(inputStream, file.toPath(), StandardCopyOption.REPLACE_EXISTING);
-        }
+    private void fixRecordDuration() throws IOException, InterruptedException {
```
Should we rename this method to `reencodeRecording`?

oussamabadr replied:
It's done.

kiview commented on modules/selenium/src/test/java/org/testcontainers/junit/ChromeRecordingWebDriverContainerTest.java:
```diff
@@ -58,6 +64,52 @@ public String getFilesystemFriendlyName() {
+
+                try( GenericContainer container = new GenericContainer<>("jrottenberg/ffmpeg:3.2-alpine38") ) {
+                    container.withStartupCheckStrategy(new OneShotStartupCheckStrategy())
+                            .withFileSystemBind(recordedFile, "/tmp/chromeTestRecord.flv", BindMode.READ_WRITE )
```
This line might be problematic in certain CI environments. `copyFiletoContainer` should be the safer method.

oussamabadr replied:
Done.

kiview commented on modules/selenium/src/test/java/org/testcontainers/junit/ChromeRecordingWebDriverContainerTest.java:
```diff
@@ -58,6 +64,52 @@ public String getFilesystemFriendlyName() {
+
+                ToStringConsumer dockerLogConsumer = new ToStringConsumer();
+
+                try( GenericContainer container = new GenericContainer<>("jrottenberg/ffmpeg:3.2-alpine38") ) {
```
In general, we like to avoid using 3rd party images as part of our CI (if we can). I assume we could use the updated `vnc-recorder` container here as well, or?

oussamabadr replied:
Done.

kiview commented on modules/selenium/src/test/java/org/testcontainers/junit/ChromeRecordingWebDriverContainerTest.java:
```diff
@@ -58,6 +64,52 @@ public String getFilesystemFriendlyName() {
+                            .withCommand("-i" ,"/tmp/chromeTestRecord.flv")
+                            .withLogConsumer(dockerLogConsumer)
+                            .start();
+                } catch (RuntimeException ex) {
```
`OneShotStartupCheckStrategy ` should not throw an exception if the container stops with exit code 0.

oussamabadr replied:
Fixed.

kiview reviewed (changes requested):
Thanks a lot, I have added some discussion point.

Looking forward to getting this quality of life fix merged 🙂

leonard84 commented on core/src/main/java/org/testcontainers/containers/VncRecordingContainer.java:
```diff
@@ -87,19 +88,27 @@ protected void configure() {
-            Files.copy(inputStream, file.toPath(), StandardCopyOption.REPLACE_EXISTING);
-        }
+    private void reencodeRecording() throws IOException, InterruptedException {
+        String newFlvOutput = "/newScreen.flv";
```
Why encode to `.flv` instead of `.mp4` container? Only to avoid the `-movflags faststart` parameter?

oussamabadr replied:
I don't know if it's Ok to change the record extension in this ticket, maybe we can open a new ticket for this, and I add that we can implement a kind of configuration to let user decide which extension he wants for recording...

...but, I'm open if you want me to change my pull request to use `.mp4` instead of `.flv`  with `-movflags faststart`  :slightly_smiling_face:

kiview replied:
I personally have no opinion about this. 
@leonard84 could you elaborate why `.mp4` would be the better choice? Would be happy to endorse this instead and change it in this PR, if it is the better choice for whatever reason.

leonard84 replied:
> FLV is a proprietary format developed by Adobe. MP4 is a format standardized by the ISO in the MPEG-4 family. Your video is H.264 (MPEG-4 Part 10), so it's from the same MPEG-4 standard.2

>That being said, the support for MP4 should always be better than for FLV containers. While typical desktop video players support almost every container and codec, devices such as mobile phones or DLNA streaming servers might only handle MP4 files.

https://superuser.com/a/549647

oussamabadr replied:
I can add this blog article of Adobe about `flv` being obsolete: 

> If you want to create a video to play in Flash Player, then you should generally be using H.264 (.mp4), not the obsolete FLV or F4V formats...


https://blogs.adobe.com/creativecloud/removal-flv-f4v-swf-export-adobe-media-encoder-after-effects-premiere-pro/

leonard84 replied:
IMHO we should offer `mp4` as default and could offer a legacy mode that produces `flv`. 
I don't think that there are any who specifically want `flv`, we currently transcode it in a post processing  step to `mp4`.
Sure this change breaks that, but that is the whole point, and we can just remove the step with this feature.

leonard84 replied:
Was this quietly resolved? I still think we should use `mp4`.
```suggestion
        String newFlvOutput = "/newScreen.mp4";
        execInContainer("ffmpeg" , "-i", RECORDING_FILE_NAME, "-vcodec", "libx264", "-movflags", "faststart", newFlvOutput);
```

Since the comment already has fixed line, I can't really let the suggestion replace the next line.

kiview commented on modules/selenium/src/test/java/org/testcontainers/junit/ChromeRecordingWebDriverContainerTest.java:
```diff
@@ -58,6 +66,52 @@ public String getFilesystemFriendlyName() {
+                String recordedFile = vncRecordingDirectory.getRoot().listFiles(new PatternFilenameFilter(flvFileNameRegEx))[0].getCanonicalPath();
+
+                String dockerOutput;
+                try( GenericContainer<?> container = new GenericContainer<>(TestcontainersConfiguration.getInstance().getVncDockerImageName()) ) {
```
I wonder, can't we use `VncRecordingContainer` here directly? You can still override `cmdModifier` etc.

oussamabadr replied:
`VncRecordingContainer` container has constructors with `@NonNull` arguments, and I don't know how can I instantiate the container without creating a method to return a `VncRecordingContainer` with `ffmpeg` as an entrypoint, which I think not good for the class as It will deviate from its main role which is  a vnc recoding, and `ffmpeg` is just an implementation detail (To fix the recording), and in my test I just want to get meta data for the recorded video using whatever `ffmpeg`.

Nonetheless, I can apply your suggestion, I just need some hints on how to do it properly, I don't want to mess the code up, as I don't have a deep knowledge of the base code  :sweat_smile:

kiview replied:
ok, makes sense.

kiview commented on modules/selenium/src/test/java/org/testcontainers/junit/ChromeRecordingWebDriverContainerTest.java:
```diff
@@ -58,6 +66,52 @@ public String getFilesystemFriendlyName() {
+
+                String recordedFile = vncRecordingDirectory.getRoot().listFiles(new PatternFilenameFilter(flvFileNameRegEx))[0].getCanonicalPath();
+
+                String dockerOutput;
```
```suggestion
                String dockerOutput = "";
```

Maybe better to initialize with empty String, so we can't end up with NPE in the assert.

oussamabadr replied:
It's no longer the case, I merged the declaration and the assignment, thanks :thumbsup:

oussamabadr commented:
Some tests doesn't pass :
`ChromeWebDriverContainerTest. simpleTest`
`CustomWaitTimeoutWebDriverContainerTest. simpleTest`
`SpecificImageNameWebDriverContainerTest. simpleTest`

due to [ExpectedConditions.visibilityOfAllElementsLocatedBy(By.cssSelector("#search h3")](https://github.com/testcontainers/testcontainers-java/blob/3ad8d80e2484864e554744a4800a81f6b7982168/modules/selenium/src/test/java/org/testcontainers/junit/BaseWebDriverContainerTest.java#L33) that is not met, but when using just `By.cssSelector("#search")` all tests pass.

I will try find out why and commit a fix

oussamabadr commented:
The tests was fixed by changing the Css Selector to look for parent element `#search` then for child elements with `h3` tag (See [last commit](https://github.com/testcontainers/testcontainers-java/pull/3180/commits/30963c2fae95272bcda09b0a6a763be143e1e730)):

`ChromeWebDriverContainerTest. simpleTest`
`CustomWaitTimeoutWebDriverContainerTest. simpleTest`
`SpecificImageNameWebDriverContainerTest. simpleTest`

rnorth commented:
Updated docker image is published: https://hub.docker.com/layers/testcontainers/vnc-recorder/1.2.0/images/sha256-1e14fc21b1dd7c28978f82f90008e4a603cc62744631cd299d3f846a08821ea9?context=repo

oussamabadr commented:
+ I have updated to use the new [testcontainers/vnc-recorder:1.2.0](https://hub.docker.com/layers/testcontainers/vnc-recorder/1.2.0/images/sha256-1e14fc21b1dd7c28978f82f90008e4a603cc62744631cd299d3f846a08821ea9?context=explore).
+ I canceled my last fix of about Css Selector due to changes in org.testcontainers.junit.BaseWebDriverContainerTest#doSimpleExplore in master (and now all tests pass).
+ I tried to switch to `mp4` but some tests was unstable, so I prefer to have a separated ticket for it (I can work on it) and also indicate whether should we make `flv` available as an option or not for backward compatibility.

kiview commented on modules/selenium/src/test/java/org/testcontainers/junit/ChromeRecordingWebDriverContainerTest.java:
```diff
@@ -59,6 +66,53 @@ public String getFilesystemFriendlyName() {
+            ) {
+                chrome.start();
+
+                TimeUnit.SECONDS.sleep(5);
```
Why do you need this sleep?

oussamabadr replied:
Sometimes the test goes very fast and the record may have **duration = 00:00:00.00**,  I reduce sleep to 1 second.

kiview commented on modules/selenium/src/test/java/org/testcontainers/junit/ChromeRecordingWebDriverContainerTest.java:
```diff
@@ -59,6 +66,53 @@ public String getFilesystemFriendlyName() {
+                            .withStartupTimeout( Duration.ofMinutes(1) )
+                            .withCommand("-i" , recordFileContainerPath, "-f" ,"null" ,"-" )
+                            .start();
+                    String ffmpegOutput = container.getLogs();
```
I think it is not guaranteed, that the command did finish, therefore you miss the logs in CI, which runs slower. This container is missing the appropriate Wait Strategy. `LogMessageWaitStrategy` seems to be a good candidate here.

oussamabadr replied:
done, thanks for your help

rnorth commented:
I've merged from master, as we've had a problem with GitHub Actions that needed to be fixed in the master branch.

kiview commented on core/src/main/java/org/testcontainers/utility/TestcontainersConfiguration.java:
```diff
@@ -105,6 +105,10 @@ public String getSocatContainerImage() {
+        return getVncDockerImageName().asCanonicalNameString();
+    }
+
+    public DockerImageName getVncDockerImageName() {
```
Maybe this is my mistake from solving the merge conflict, but why is `getVncDockerImageName()` needed?

Can't you use `getVncRecordedContainerImage()`? (which I admit, has probably a typo, the name is strange)

oussamabadr replied:
Done, thanks.

kiview commented on modules/selenium/src/test/java/org/testcontainers/junit/ChromeRecordingWebDriverContainerTest.java:
```diff
@@ -59,6 +68,55 @@ public String getFilesystemFriendlyName() {
+        public void recordingTestThatShouldHaveCorrectDuration() throws IOException, InterruptedException {
+            File target = vncRecordingDirectory.getRoot();
+            try (
+                // recordAll {
+                // To do this, simply add extra parameters to the rule constructor:
```
You can remove those comments. They are our markers for generating code includes in the docs.

```suggestion
```

oussamabadr replied:
Done, thanks.

kiview commented on modules/selenium/src/test/java/org/testcontainers/junit/ChromeRecordingWebDriverContainerTest.java:
```diff
@@ -59,6 +68,55 @@ public String getFilesystemFriendlyName() {
+            ) {
+                chrome.start();
+
+                TimeUnit.SECONDS.sleep(1);
```
Why sleep? Shouldn't we get a duration > 0 either way?

oussamabadr replied:
Sometimes the test goes very fast and the record may have `duration = 00:00:00.00`, I reduced it to 500 milliseconds.

kiview commented on modules/selenium/src/test/java/org/testcontainers/junit/ChromeRecordingWebDriverContainerTest.java:
```diff
@@ -59,6 +68,55 @@ public String getFilesystemFriendlyName() {
+                    }
+                }, Optional.empty());
+
+                String flvFileTitle = "ChromeThatRecordsAllTests-recordingTestThatShouldBeRecordedAndRetained";
```
Better to have a unique name here and also use this in the regex below (this was copied from the test above).

oussamabadr replied:
Done, thanks.

kiview reviewed (changes requested):
Overall looks good, just had some small remarks.

@rnorth 
Can you have a final look?

kiview commented on modules/selenium/src/test/java/org/testcontainers/junit/ChromeRecordingWebDriverContainerTest.java:
```diff
@@ -59,6 +68,54 @@ public String getFilesystemFriendlyName() {
+            ) {
+                chrome.start();
+
+                TimeUnit.MILLISECONDS.sleep(500);
```
```suggestion
                TimeUnit.MILLISECONDS.sleep(500); // to make sure video will have a duration
```

kiview reviewed (approved):
@rnorth Can you have a final look?

leonard84 commented on core/src/main/java/org/testcontainers/containers/VncRecordingContainer.java:
```diff
@@ -87,19 +88,27 @@ protected void configure() {
-            Files.copy(inputStream, file.toPath(), StandardCopyOption.REPLACE_EXISTING);
-        }
+    private void reencodeRecording() throws IOException, InterruptedException {
+        String newFlvOutput = "/newScreen.flv";
+        execInContainer("ffmpeg" , "-i", RECORDING_FILE_NAME, "-vcodec", "libx264", newFlvOutput);
```
Since commenting doesn't seem to reopen the issue, I have to make a new one https://github.com/testcontainers/testcontainers-java/pull/3180/files#r524394794

```suggestion
        String newFlvOutput = "/newScreen.mp4";
        execInContainer("ffmpeg" , "-i", RECORDING_FILE_NAME, "-vcodec", "libx264", "-movflags", "faststart", newFlvOutput);
```

rnorth reviewed (changes requested):
We had a discussion offline, and decided that we can't risk changing the default emitted file format without warning.

I'll take an action to amend this PR, adding a parameter for format which will default to FLV. In 1.16.0 we'll change the default to MP4.

Sorry that this came as a late decision.

leonard84 commented:
@rnorth sounds good.

> I'll take an action to amend this PR, adding a parameter for format which will default to FLV. In 1.16.0 we'll change the default to MP4.

Just note that it does not suffice to change the extension `mp4` requires `-movflags faststart` to create stream able video.

oussamabadr commented:
@rnorth @kiview  I have added `MP4` support with `FLV` as default.
I have also changed [ChromeRecordingWebDriverContainerTest](https://github.com/testcontainers/testcontainers-java/blob/8b78a9ce4ea9a90661935e5db6d87aa26b858fe0/modules/selenium/src/test/java/org/testcontainers/junit/ChromeRecordingWebDriverContainerTest.java), the tests required minimum duration to correctly process video with `ffmpeg` (I'm still in doubt for this one)

I also updated documentation by mentioning `mp4` support.

@leonard84 I used `-movflags faststart` options for `mp4`.

Please let me know if I missed something. 

Thanks.

leonard84 commented on core/src/main/java/org/testcontainers/containers/VncRecordingContainer.java:
```diff
@@ -97,18 +106,44 @@ public void saveRecordingToFile(File file) {
+    public enum VncRecordingFormat {
+        FLV("flv") {
+            @Override
+            String reencodeRecording(VncRecordingContainer container, String source) throws IOException, InterruptedException {
```
What about `@NonNull` for the parameters

oussamabadr replied:
Done, Thanks

leonard84 commented on modules/selenium/src/main/java/org/testcontainers/containers/RecordingFileFactory.java:
```diff
@@ -1,15 +1,16 @@
     @Deprecated
-    default File recordingFileForTest(File vncRecordingDirectory, Description description, boolean succeeded) {
-        return recordingFileForTest(vncRecordingDirectory, description.getTestClass().getSimpleName() + "-" + description.getMethodName(), succeeded);
+    default File recordingFileForTest(File vncRecordingDirectory, Description description, boolean succeeded, VncRecordingFormat recordingFormat) {
```
I would not break existing interfaces and update this method

oussamabadr replied:
Done, I canceled the changes.
Thanks

leonard84 commented on modules/selenium/src/main/java/org/testcontainers/containers/RecordingFileFactory.java:
```diff
@@ -1,15 +1,16 @@
     }
 
-    File recordingFileForTest(File vncRecordingDirectory, String prefix, boolean succeeded);
+    File recordingFileForTest(File vncRecordingDirectory, String prefix, boolean succeeded, VncRecordingFormat recordingFormat);
```
This breaks existing implementations, I'm not sure what the most elegant solution would be.

This would create a stackoverflow if you don't implement any of the new methods, but if you already have one that is using the three param `recordingFileForTest` it would work. 
```java
@Deprecated
default File recordingFileForTest(File vncRecordingDirectory, String prefix, boolean succeeded) { 
    return recordingFileForTest(vncRecordingDirectory, prefix, boolean, null);
}
    default File recordingFileForTest(File vncRecordingDirectory, String prefix, boolean succeeded, VncRecordingFormat recordingFormat) {
    return recordingFileForTest(vncRecordingDirectory, prefix, boolean);
}
```
The alternative would be to just `return null` from the three param default method, since our code would call the four param version. In any case we would need to add default methods so that users are not forced to implement an unused method.

oussamabadr replied:
Indeed, I used `Default` for the new method, so user can implement it only if `mp4` is desired, or setting `flv` explicitly.
I canceled the changes in the three params method, I didn't use `Deprecated` and I considered It as a legitimate method (recording with default format).

Thanks.

leonard84 commented on core/src/main/java/org/testcontainers/containers/VncRecordingContainer.java:
```diff
@@ -83,23 +93,57 @@ protected void configure() {
     }
 
+    @SneakyThrows
+    public void saveRecordingToFile(File file) {
```
Add `@NonNull `

oussamabadr replied:
Done, Thanks

leonard84 commented on modules/selenium/src/main/java/org/testcontainers/containers/BrowserWebDriverContainer.java:
```diff
@@ -358,8 +361,13 @@ public SELF withLinkToContainer(LinkableContainer otherContainer, String alias)
+        return withRecordingMode(recordingMode, vncRecordingDirectory, null);
+    }
+
+    public SELF withRecordingMode(VncRecordingMode recordingMode, File vncRecordingDirectory, VncRecordingFormat recordingFormat) {
```
@NonNull should also be added where it applies

oussamabadr replied:
I used `@Nullable` instead for `recordingFormat`.  
thanks.

leonard84 reviewed (changes requested):
Looks good, just some small issues

leonard84 commented:
@kiview / @rnorth could you take another look.

oussamabadr commented:
I have just updated with the latest commit from **master**.
could you please have a look ? (so it can be soon merged to the main branch).

Thank you

bsideup commented on core/src/main/java/org/testcontainers/containers/VncRecordingContainer.java:
```diff
@@ -71,6 +74,13 @@ public VncRecordingContainer withVncPort(int vncPort) {
         return this;
     }
 
+    public VncRecordingContainer withVideoExtension(VncRecordingFormat recordingFormat) {
```
shouldn't it be `withVideoFormat`?

oussamabadr replied:
Yes, I think it's a better name, Done.

bsideup commented on core/src/main/java/org/testcontainers/containers/VncRecordingContainer.java:
```diff
@@ -83,23 +93,69 @@ protected void configure() {
 
+    @SneakyThrows
+    public void saveRecordingToFile(@NonNull File file) {
+        try(InputStream inputStream = streamRecording()) {
```
```suggestion
        try (InputStream inputStream = streamRecording()) {
```

oussamabadr replied:
Done

bsideup commented on core/src/main/java/org/testcontainers/containers/VncRecordingContainer.java:
```diff
@@ -83,23 +93,69 @@ protected void configure() {
+        /**
+         * @return {@code vncRecordingFormat} value if not null, otherwise, {@link VncRecordingFormat#FLV} will be returned as a default format.
+         */
+        public static VncRecordingFormat of(VncRecordingFormat vncRecordingFormat) {
```
what's the purpose of this method?

oussamabadr replied:
Just in case of `null`, so It can fall back to the default format. Indeed I wasn't sure If it was necessary, but just in case, to prevent a possible `NullPointerException`. 
But I get rid of it for now, thanks.

bsideup commented on modules/selenium/src/main/java/org/testcontainers/containers/BrowserWebDriverContainer.java:
```diff
@@ -358,8 +361,13 @@ public SELF withLinkToContainer(LinkableContainer otherContainer, String alias)
+        return withRecordingMode(recordingMode, vncRecordingDirectory, null);
+    }
+
+    public SELF withRecordingMode(VncRecordingMode recordingMode, File vncRecordingDirectory, @Nullable VncRecordingFormat recordingFormat) {
```
Let's not make it Nullable:

```suggestion
    public SELF withRecordingMode(VncRecordingMode recordingMode, File vncRecordingDirectory, VncRecordingFormat recordingFormat) {
```

oussamabadr replied:
Done

bsideup commented on modules/selenium/src/main/java/org/testcontainers/containers/DefaultRecordingFileFactory.java:
```diff
@@ -4,20 +4,29 @@
             prefix,
-            filenameDateFormat.format(new Date())
+            filenameDateFormat.format(new Date()),
+            VncRecordingFormat.of(recordingFormat).getExtension()
```
Should not be `null`:
```suggestion
            recordingFormat.getExtension()
```

oussamabadr replied:
Done

bsideup commented on modules/selenium/src/main/java/org/testcontainers/containers/DefaultRecordingFileFactory.java:
```diff
@@ -4,20 +4,29 @@
 
     @Override
     public File recordingFileForTest(File vncRecordingDirectory, String prefix, boolean succeeded) {
+        return recordingFileForTest(vncRecordingDirectory, prefix, succeeded, null);
```
```suggestion
        return recordingFileForTest(vncRecordingDirectory, prefix, succeeded, VncRecordingFormat.FLV);
```

oussamabadr replied:
Done, but with a little difference, tell me if it's ok for you.

bsideup commented on modules/selenium/src/test/java/org/testcontainers/junit/ChromeRecordingWebDriverContainerTest.java:
```diff
@@ -10,55 +10,143 @@
+                mountableFile = MountableFile.forHostPath(recordedFiles[0].getCanonicalPath());
+            }
+
+            try( GenericContainer<?> container = new GenericContainer<>(DockerImageName.parse("testcontainers/vnc-recorder:1.2.0")) ) {
```
```suggestion
            try (GenericContainer<?> container = new GenericContainer<>(DockerImageName.parse("testcontainers/vnc-recorder:1.2.0"))) {
```

oussamabadr replied:
Done

bsideup commented on modules/selenium/src/test/java/org/testcontainers/junit/ChromeRecordingWebDriverContainerTest.java:
```diff
@@ -10,55 +10,143 @@
+                container.withCopyFileToContainer(mountableFile, recordFileContainerPath)
+                        .withCreateContainerCmdModifier( createContainerCmd -> createContainerCmd.withEntrypoint("ffmpeg") )
+                        .withCommand("-i" , recordFileContainerPath, "-f" ,"null" ,"-" )
+                        .waitingFor( new LogMessageWaitStrategy()
```
```suggestion
                        .waitingFor(new LogMessageWaitStrategy()
```

oussamabadr replied:
Done

bsideup commented on modules/selenium/src/test/java/org/testcontainers/junit/ChromeRecordingWebDriverContainerTest.java:
```diff
@@ -10,55 +10,143 @@
+                String recordFileContainerPath = "/tmp/chromeTestRecord.flv";
+                container.withCopyFileToContainer(mountableFile, recordFileContainerPath)
+                        .withCreateContainerCmdModifier( createContainerCmd -> createContainerCmd.withEntrypoint("ffmpeg") )
+                        .withCommand("-i" , recordFileContainerPath, "-f" ,"null" ,"-" )
```
```suggestion
                        .withCommand("-i", recordFileContainerPath, "-f" ,"null" ,"-" )
```

oussamabadr replied:
Done

bsideup commented on modules/selenium/src/test/java/org/testcontainers/junit/ChromeRecordingWebDriverContainerTest.java:
```diff
@@ -10,55 +10,143 @@
+            try( GenericContainer<?> container = new GenericContainer<>(DockerImageName.parse("testcontainers/vnc-recorder:1.2.0")) ) {
+                String recordFileContainerPath = "/tmp/chromeTestRecord.flv";
+                container.withCopyFileToContainer(mountableFile, recordFileContainerPath)
+                        .withCreateContainerCmdModifier( createContainerCmd -> createContainerCmd.withEntrypoint("ffmpeg") )
```
```suggestion
                        .withCreateContainerCmdModifier(createContainerCmd -> createContainerCmd.withEntrypoint("ffmpeg"))
```

oussamabadr replied:
Done

oussamabadr commented:
Thanks everyone for letting me contribute to this awesome project, hope it will be useful to the community  :smiley:

bsideup commented:
@oussamabadr thank you so much for contributing it! 💯 
The timing was a bit uneasy given the reduced capacity of the team, and I am very sorry it took so long, but we're very happy with the result!

Expect it to be released very soon :)

kayb-proxora commented:
Hi,

For me the screen recording is no longer working after this change. It seems that ffmpeg does not produce any output.

Can someone please check?

Thanks and kind regards,

Kay

kiview commented:
@AB-xdev 
This PR was released in 1.15.2.

The default is FLV, but you can specify MP4 as well.

AB-xdev commented:
@kiview 
> This PR was released in 1.15.2.
> 
> The default is FLV, but you can specify MP4 as well.

Thank you for the ultra-fast feedback 👍 
However I meant the change that the default format is mp4 (or at least it should be as stated in the comment above).

Backstory:
We have an (non critical) issue in our project that basically says "Change Testconatiner-Test videos from FLV to MP4". 
However we read that you will change it anyways with 1.16.0 and thought that we can simply sit this out and fix it on the fly when we update to 1.16 😄

bsideup commented:
@AB-xdev we forgot to switch the default in 1.16.0, sorry 😅

bsideup commented:
@AB-xdev we do! It is the right default value and will make many users' lives easier, just we need to do it in a "major" release, to not confuse our users. So, 1.17.0 then 😅
